lib/grandpa/tracker_commits.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"container/list"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+// commitsTracker tracks vote messages that could
+// not be processed, and removes the oldest ones once
+// its maximum capacity is reached.
+// It is NOT THREAD SAFE to use.
+type commitsTracker struct {
+	// map of commit block hash to data
+	// data = message + tracking linked list element pointer
+	mapping map[common.Hash]commitMessageMapData
+	// double linked list of block hash
+	// to track the order commit messages were added in.
+	linkedList *list.List
+	capacity   int
+}
+
+type commitMessageMapData struct {
+	message *CommitMessage
+	// element contains a block hash value.
+	element *list.Element
+}
+
+// newCommitsTracker creates a new commit messages tracker
+// with the capacity specified.
+func newCommitsTracker(capacity int) commitsTracker {
+	return commitsTracker{
+		mapping:    make(map[common.Hash]commitMessageMapData, capacity),
+		linkedList: list.New(),
+		capacity:   capacity,
+	}
+}
+
+// add adds a commit message to the commit message tracker.
+// If the commit message tracker capacity is reached,
+// the oldest commit message is removed.
+func (ct *commitsTracker) add(commitMessage *CommitMessage) {
+	blockHash := commitMessage.Vote.Hash
+
+	data, has := ct.mapping[blockHash]
+	if has {
+		// commit already exists so override the commit for the block hash;
+		// do not move the list element in the linked list to avoid
+		// someone re-sending the same commit message and going at the
+		// front of the list, hence erasing other possible valid commit messages
+		// in the tracker.
+		data.message = commitMessage
+		ct.mapping[blockHash] = data
+		return
+	}
+
+	// add new block hash in tracker
+	ct.cleanup()
+	element := ct.linkedList.PushFront(blockHash)
+	data = commitMessageMapData{
+		message: commitMessage,
+		element: element,
+	}
+	ct.mapping[blockHash] = data
```

Comment:
not sure why we need the `commitMessageMapData` type? Could we not just use `list.PushFront` to return us an element and store that in `ct.mapping`? That way `ct.mapping` would be of type `map[common.Hash]*list.Element`

Reply:
You would have a pointer to the `list.Element` and you could just use the `list.Element.Value` attribute and cast it back to `*CommitMessage`. 

I took a stab at it on another branch: https://github.com/ChainSafe/gossamer/pull/2517/files

Reply:
I think it's a minor optimisation.  I think avoiding more types is probably best.